NDTensors/src/lib/Sectors/src/abstractcategory.jl

```diff
@@ -1,45 +1,143 @@
+istrivial(c::AbstractCategory) = (c == trivial(typeof(c)))
+
+function category_label(c::AbstractCategory)
+  return error("method `category_label` not defined for type $(typeof(c))")
+end
+
+function GradedAxes.dual(category_type::Type{<:AbstractCategory})
+  return error("`dual` not defined for type $(category_type).")
+end
+
+block_boundaries(g::AbstractUnitRange) = block_boundaries(SymmetryStyle(g), g)
+block_boundaries(::AbelianGroup, g) = GradedAxes.unlabel.(BlockArrays.blocklengths(g))
+function block_boundaries(::NonAbelianGroup, g)
+  return Sectors.quantum_dimension.(GradedAxes.blocklabels(g)) .*
+         BlockArrays.blocklengths(g)
+end
+
+quantum_dimension(x) = quantum_dimension(SymmetryStyle(x), x)
+
+function quantum_dimension(::SymmetryStyle, c::AbstractCategory)
+  return error("method `quantum_dimension` not defined for type $(typeof(c))")
 end
 
-function fusion_rule(c1::AbstractCategory, c2::AbstractCategory)
-  category_type = typeof(c1)
-  return [category_type(l) for l in label_fusion_rule(category_type, label(c1), label(c2))]
+quantum_dimension(::AbelianGroup, ::AbstractCategory) = 1
+quantum_dimension(::EmptyCategory, ::AbstractCategory) = 0
+
+function quantum_dimension(::SymmetryStyle, g::AbstractUnitRange)
+  gblocks = BlockArrays.blocklengths(g)
+  return sum(gblocks .* quantum_dimension.(LabelledNumbers.label.(gblocks)))
 end
 
+quantum_dimension(::AbelianGroup, g::AbstractUnitRange) = length(g)
+function quantum_dimension(::SymmetryStyle, g::GradedAxes.UnitRangeDual)
+  return quantum_dimension(GradedAxes.dual(g))
+end
+quantum_dimension(::AbelianGroup, g::GradedAxes.UnitRangeDual) = length(g)  # resolves ambiguity
+
+# ================  fusion rule interface ====================
 ⊗(c1::AbstractCategory, c2::AbstractCategory) = fusion_rule(c1, c2)
 
-⊕(c1::AbstractCategory, c2::AbstractCategory) = [c1, c2]
-⊕(cs::Vector{<:AbstractCategory}, c::AbstractCategory) = [cs; c]
-⊕(c::AbstractCategory, cs::Vector{<:AbstractCategory}) = [c; cs]
+function fusion_rule(c1, c2)
+  return fusion_rule(combine_styles(SymmetryStyle(c1), SymmetryStyle(c2)), c1, c2)
+end
 
-function Base.show(io::IO, cs::Vector{<:AbstractCategory})
-  (length(cs) <= 1) && print(io, "[")
-  symbol = ""
-  for c in cs
-    print(io, symbol, c)
-    symbol = " ⊕ "
-  end
-  (length(cs) <= 1) && print(io, "]")
-  return nothing
+function fusion_rule(::SymmetryStyle, c1::C, c2::C) where {C<:AbstractCategory}
+  degen, labels = label_fusion_rule(C, category_label(c1), category_label(c2))
+  return GradedAxes.gradedrange(LabelledNumbers.LabelledInteger.(degen, C.(labels)))
 end
 
-function trivial(category_type::Type{<:AbstractCategory})
-  return error("`trivial` not defined for type $(category_type).")
+# abelian case: return Category
+function fusion_rule(::AbelianGroup, c1::C, c2::C) where {C<:AbstractCategory}
+  return C(label_fusion_rule(C, category_label(c1), category_label(c2)))
```

Comment:
```suggestion
  return label_fusion_rule(C, category_label(c1), category_label(c2))
```
Was the explicit conversion to `C` needed?

Reply:
I think the conversion shouldn't be in the generic version `fusion_rule(::AbelianGroup, c1::C, c2::C) where {C<:AbstractCategory}` but instead in the specific `label_fusion_rule` definitions as needed, i.e.:
```julia
label_fusion_rule(::Type{Ising}, l1::C, l2::C) where {C<:AbstractCategory} = C(label_fusion_rule(su2{2}, l1, l2))
```

Reply:
Oh sorry, I was was misinterpreting the code here, I see that wrapping the result in `C(...)` here:
```julia
function fusion_rule(::AbelianGroup, c1::C, c2::C) where {C<:AbstractCategory}
  return C(label_fusion_rule(C, category_label(c1), category_label(c2)))
end
```
is not just converting the output but rewrapping the category label back into the category, I was losing track of `fusion_rule` and `label_fusion_rule`. I think the code makes sense as it is.